allensdk/brain_observatory/behavior/behavior_project_cache/tables/util/metadata_parsers.py

```diff
@@ -0,0 +1,245 @@
+
+    if session_type not in SESSION_TYPES:
+        return None
+    is_passive = (
+        any(
+            passive_num in session_type
+            for passive_num in passive_ophys_session_nums
+        )
+        and "OPHYS" in session_type
+    )
+    if is_passive:
+        behavior_type = "passive_viewing"
+    else:
+        behavior_type = "active_behavior"
+    return behavior_type
+
+
+def parse_stimulus_set(metadata_row: pd.Series) -> Union[str, None]:
+    """Return the name of the image set or gratings.
+
+    Parameters
+    ----------
+    metadata_row : pandas.Series
+        Complete set of metadata for session.
+
+    Returns
+    -------
+    stim_set : str
+        Name of stimulus set shown for session.
+    """
+    session_type = metadata_row["session_type"]
+    if session_type not in SESSION_TYPES:
+        return None
+
+    split_str = session_type.split("_")
+    if split_str[2] == "images":
+        stim_set = split_str[2] + "_" + split_str[3]
+    elif split_str[2] == "gratings":
+        stim_set = "gratings"
+    return stim_set
+
+
+def parse_stimulus_set_novelty(metadata_row: pd.Series) -> Union[str, None]:
+    """Return familiarity with stimulus for session.
+
+    Parameters
+    ----------
+    metadata_row : pandas.Series
+        Complete set of metadata for session.
+
+    Returns
+    -------
+    stimulus_set_novelty : str
+        ``familiar`` or ``novel`` for session.
+    """
+    familiar_ophys_sessions = ["0", "1", "2", "3"]
+
+    session_type = metadata_row["session_type"]
+    if session_type not in SESSION_TYPES:
+        return None
+
+    if "TRAINING" in session_type:
+        stimulus_set_novelty = "training"
+    elif "7" in session_type:
+        stimulus_set_novelty = "NA"
+    else:
+        is_familiar = any(
+            familiar_num in session_type
+            for familiar_num in familiar_ophys_sessions
+        )
+        if is_familiar:
+            stimulus_set_novelty = "familiar"
+        else:
+            stimulus_set_novelty = "novel"
+    return stimulus_set_novelty
+
+
+############################
+#
+#     from project code
+#
+############################
+
+
+def parse_num_cortical_structures(metadata_row: pd.Series) -> Union[str, None]:
+    """Return the number of structures that were targeted for this session.
+
+    Parameters
+    ----------
+    metadata_row : pandas.Series
+        Complete set of metadata for session.
+
+    Returns
+    -------
+    num_structures : int
+        Number of structures targeted for the session.
+    """
+    attr_names = dir(metadata_row)
+    if "project_code" in attr_names:
+        project_code = metadata_row["project_code"]
+    elif "project_code_behavior" in attr_names:
+        project_code = metadata_row["project_code_behavior"]
+    elif "project_code_ophys" in attr_names:
+        project_code = metadata_row["project_code_ophys"]
+    else:
+        return None
+
+    if project_code not in PROJECT_CODES:
```

Comment:
don't need to check if do `NUM_STRUCTURES_DICT.get(project_code, None)`
I also prefer to raise exception rather than silently returning None

Reply:
Will switch to the simpler code. See above comment regarding returning None.